apollo-backend/src/main/java/io/logz/apollo/deployment/DeploymentService.java

```diff
@@ -0,0 +1,62 @@
+package io.logz.apollo.deployment;
+
+import io.logz.apollo.dao.DeploymentDao;
+import io.logz.apollo.models.Deployment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@Singleton
+public class DeploymentService {
+    private static final Logger logger = LoggerFactory.getLogger(DeploymentService.class);
+    private final Duration RUNNING_DEPLOYMENT_TIMEOUT = Duration.ofHours(12);
+
+    private final DeploymentDao deploymentDao;
+    private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+
+    @Inject
+    public DeploymentService(DeploymentDao deploymentDao) {
+        this.deploymentDao = deploymentDao;
+    }
+
+    @PostConstruct
+    private void scheduleTask() {
+        executor.scheduleWithFixedDelay(this::cancelStuckDeployments, 0L, 30, TimeUnit.MINUTES);
+    }
+
+    private void cancelStuckDeployments() {
+        try {
+            List<Integer> stuckDeployments = deploymentDao.getAllRunningDeployments().stream()
+                    .filter(deployment -> deployment.getLastUpdate().toInstant().isBefore(Instant.now().minus(RUNNING_DEPLOYMENT_TIMEOUT)))
+                    .map(Deployment::getId)
+                    .collect(Collectors.toList());
+
+            logger.warn("Found {} stuck deployments, cancelling the following ids: {}", stuckDeployments.size(), stuckDeployments);
+            stuckDeployments.forEach(deploymentId -> deploymentDao.updateDeploymentStatus(deploymentId, Deployment.DeploymentStatus.CANCELED));
```

Comment:
Not sure about this.
You can't go from any state to canceled, this can cause big issues.
I'm ok with moving everything that is stuck in "Canceling" to "Canceled", but not moving  "Pending"  or "Started" directly to this

Reply:
Most of the issue are with "pending" which get stuck and then @mayhuberman needs to manually change it to "cancelled" in the db manually. So we do the same here.
"Started" could be an issue, so maybe i'll change it to only pending or cancelling.